odl/solvers/advanced/douglas_rachford.py

```diff
@@ -0,0 +1,168 @@
+# Copyright 2014-2016 The ODL development group
+#
+# This file is part of ODL.
+#
+# ODL is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# ODL is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+""" Implementation of the Douglas-Rachford splitting algorithm """
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from future import standard_library
+standard_library.install_aliases()
+
+from odl.operator import Operator
+
+
+__all__ = ('douglas_rachford_pd',)
+
+
+def douglas_rachford_pd(x, prox_f, prox_cc_g, L, tau, sigma, niter,
+                        callback=None, **kwargs):
+    """Douglas-Rachford primal-dual splitting algorithm.
+
+    Minimizes the sum of several convex functions composed with linear
+    operators
+
+        ``min_x f(x) + sum_i g_i(L_i x)``
+
+    Where f, g_i are convex functions, L_i are linear `Operator`'s.
+
+    Can also be used to solve the more general problem
+
+        ``min_x f(x) + sum_i (g_i @ l_i)(L_i x)``
+
+    Where l_i are convex functions and  @ is the infimal convolution:
+
+        ``(f @ g)(x) = inf_y { f(x-y) + g(y) }``
+
+    Parameters
+    ----------
+    x : `LinearSpaceVector`
+        Initial point, updated in place.
+    prox_f : `callable`
+        Function returning an `Operator` when called with stepsize.
+        The Operator should be the proximal of ``f``.
```

Comment:
This could e.g. be written as 
```
`Proximal factory` for the functional ``f``.
```
And then a glossary term "proximal factory". We should also point to a location where the prox factories can be found. That hint could be even before `Parameters`.